### PR TITLE
Refactor Requirements to use a BranchProtectionRequest

### PR DESCRIPTION
**Pull request by fejta**
/assign @stevekuznetsov @BenTheElder @cjwagner

### CODE REVIEW

File: prow/cmd/branchprotector/protect.go [hunk 1]
```diff
@@ -170,7 +163,7 @@ func (p *Protector) Protect() {
 
 	// Scan the branch-protection configuration
 	for orgName, org := range bp.Orgs {
-		if err := p.UpdateOrg(orgName, org, bp.Protect); err != nil {
+		if err := p.UpdateOrg(orgName, org, bp.Protect != nil); err != nil {
```
Comment (stevekuznetsov) [inline prow/cmd/branchprotector/protect.go]: I don't really understand this cascading `*bool` pass -- why do we try to resolve this at the call-sites for all of the `p.Update*` methods? The `(c *Config) GetBranchProtection` call should be able to resolve in one place for every `(org,repo,branch)` tuple whether or not that specific triad of identifiers requires protection based on explicit config and the parent-child relationships there. I see that with this approach we only conditionally call the GitHub API to get org/repo/branch lists but with the cache that should be basically free as those items really do not change often. Maybe it would be better just to get all the info and filter in one place, not in both?

File: prow/cmd/branchprotector/protect.go [hunk 2]
```diff
@@ -191,21 +184,17 @@ func (p *Protector) Protect() {
 		}
```
Comment (stevekuznetsov) [inline prow/cmd/branchprotector/protect.go]: I still feel like you should have done the config validation at load-time. This run-time checking for valid config input just makes this longer and harder to follow and we should be able to validate it well before it is used.

File: prow/cmd/branchprotector/protect.go (hunk 1 above)
Comment (fejta) [reply prow/cmd/branchprotector/protect.go]: This is passing a `bool`. The `*bool` was silly as we only check whether its nil throughout its lifetime (and then recompute the value in `GetBranchProtection`), and yes we use it exclusively to determine whether we need to read all branches/repos or just defined ones.

The cache doesn't exist yet, does it? It also feels wrong to me to read every branch in the org just because I want to protect a couple branches in a couple repos.

Would you be happier if I do not change the signature of `Update*` methods (so keep the `*bool`? The PR is fine without it.

File: prow/cmd/branchprotector/protect.go (hunk 2 above)
Comment (fejta) [reply prow/cmd/branchprotector/protect.go]: Would you be happier if I make this a panic and/or add a comment noting that this code should never run?

Given that we're doing strings.Split() there's technically no guarantee that we'll have 2 parts, and this message will be easier to debug than an index out of bounds error on the slice on L193.

File: prow/cmd/branchprotector/protect.go (hunk 1 above)
Comment (stevekuznetsov) [reply prow/cmd/branchprotector/protect.go]: Yeah I guess if we really want to optimize for GitHub API reads here this approach is fine. Was just a bit hard to follow at first.

File: prow/cmd/branchprotector/protect.go (hunk 2 above)
Comment (stevekuznetsov) [reply prow/cmd/branchprotector/protect.go]: I feel like this is just another case of where we need either:

 - users to have worse (???) UX and provide us with appropriately separated data instead of flat strings
 - internal and external config types, where the internal one here would have a struct and the type system would give us confidence in it's correctness

You're right, slice OOB would be annoying. Until we do one of the above we can keep this as-is.
